File: solutions/day_05/advent5.2.rb
#!/usr/bin/env ruby

def program
  @program ||= (
    lines = STDIN.read.split("\n").map(&:to_i)
    Hash[[(0..lines.count-1).to_a, lines].transpose]
  )
end

def execute
  step = pos = 0
  len  = program.keys.length
  while(pos < len && pos >= 0)
    val = program[pos]
    program[pos] = program[pos] + (val >= 3 ? -1 : 1)
    pos += val
    step += 1
  end
  step
end

if __FILE__ == $0
  p execute
end

File: solutions/day_05/advent5.2.ex
#!/usr/bin/env elixir

defmodule Jump do
  def program do
    IO.read(:stdio,:all)
    |> String.split
    |> Enum.map(&(String.to_integer(&1)))
    |> Stream.with_index
    |> Enum.map(fn {v,i} -> {i, v} end)
    |> Map.new
  end

  def execute(program, pos \\ 0, step \\ 0)
  def execute(program, pos, step) do
    case Map.fetch(program, pos) do
      {:ok, val} ->
        program
        |> Map.update(pos, 0, &(
            cond do
              &1 > 2 -> &1 - 1
              &1 < 3 -> &1 + 1
            end
          )
        )
        |> execute(pos + val, step + 1)
      :error -> step
    end
  end
end

Jump.program
|> Jump.execute
|> IO.inspect